src/main/cpp/src/hive_hash.cu

```diff
@@ -210,11 +208,219 @@ class hive_device_row_hasher {
+     * @param col The column device view.
+     * @param row_index The index of the row to compute the hash for.
+     * @return The computed hive hash value.
+     *
+     * @note This function is only enabled for nested column types.
+     */
     template <typename T, CUDF_ENABLE_IF(cudf::is_nested<T>())>
     __device__ hive_hash_value_t operator()(cudf::column_device_view const& col,
                                             cudf::size_type row_index) const noexcept
     {
-      CUDF_UNREACHABLE("Nested type is not supported");
+      cudf::column_device_view curr_col = col.slice(row_index, 1);
+      // The default constructor of `col_stack_element` is deleted, so it can not allocate an array
+      // of `col_stack_element` directly.
+      // Instead leverage the byte array to create the col_stack_element array.
+      alignas(col_stack_element) char stack_wrapper[sizeof(col_stack_element) * MAX_NESTED_DEPTH];
+      col_stack_element_ptr col_stack = reinterpret_cast<col_stack_element_ptr>(stack_wrapper);
+      int stack_size                  = 0;
+
+      col_stack[stack_size++] = col_stack_element(curr_col);
+
+      while (stack_size > 0) {
+        col_stack_element& element = col_stack[stack_size - 1];
+        curr_col                   = element.get_column();
+        // Do not pop it until it is processed. The definition of `processed` is:
+        // - For nested types, it is when the children are processed.
+        // - For primitive types, it is when the hash value is computed.
+        if (curr_col.type().id() == cudf::type_id::STRUCT) {
+          // All child columns processed, pop the element and update `cur_hash` of its parent column
+          if (element.cur_child_idx() == curr_col.num_child_columns()) {
+            if (--stack_size > 0) { col_stack[stack_size - 1].update_cur_hash(element.get_hash()); }
+          } else {
+            // Push the next child column into the stack
+            col_stack[stack_size++] =
+              col_stack_element(cudf::detail::structs_column_device_view(curr_col).get_sliced_child(
+                element.get_and_inc_child_idx()));
+          }
+        } else if (curr_col.type().id() == cudf::type_id::LIST) {
+          // lists_column_device_view has a different interface from structs_column_device_view
+          curr_col = cudf::detail::lists_column_device_view(curr_col).get_sliced_child();
+          if (element.cur_child_idx() == curr_col.size()) {
+            if (--stack_size > 0) { col_stack[stack_size - 1].update_cur_hash(element.get_hash()); }
+          } else {
+            col_stack[stack_size++] =
+              col_stack_element(curr_col.slice(element.get_and_inc_child_idx(), 1));
+          }
+        } else {
+          // There is only one element in the column for primitive types
+          auto hash = cudf::type_dispatcher<cudf::experimental::dispatch_void_if_nested>(
+            curr_col.type(), this->hash_functor, curr_col, 0);
```

Comment:
```suggestion
          auto hash = cudf::type_dispatcher<cudf::experimental::dispatch_void_if_nested>(
            curr_col.type(), this->hash_functor, curr_col, row_index);
```

Reply:
Here, `curr_col` has only one element, so the index should be 0.